File: src/sc_neurocore/accel/go/services/ltc.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go service for ltc

package services

import (
	"math"
)

// LiquidTimeConstantNeuronState holds the neuron state
type LiquidTimeConstantNeuronState struct {
	X float64
	TauBase float64
	WTau float64
	WX float64
	WIn float64
	Bias float64
	VThreshold float64
	Dt float64
}

// NewLiquidTimeConstantNeuron creates a new LiquidTimeConstantNeuron neuron with default parameters
func NewLiquidTimeConstantNeuron() *LiquidTimeConstantNeuronState {
	return &LiquidTimeConstantNeuronState{
		X: 0.0,
		TauBase: 10.0,
		WTau: -0.5,
		WX: 0.8,
		WIn: 1.0,
		Bias: 0.0,
		VThreshold: 1.0,
		Dt: 1.0,
	}
}

// Step advances the neuron by one timestep
func (s *LiquidTimeConstantNeuronState) Step(iExt float64) int {
	_ = iExt
	return 0
}

// SimulateLiquidTimeConstantNeuron runs the neuron for n steps
func SimulateLiquidTimeConstantNeuron(nSteps int, iExt float64) ([]float64, int) {
	s := NewLiquidTimeConstantNeuron()
	trace := make([]float64, nSteps)
	spikes := 0
	for t := 0; t < nSteps; t++ {
		result := s.Step(iExt)
		trace[t] = s.X
		if result > 0 {
			spikes++
		}
	}
	return trace, spikes
}

var _ = math.Exp

File: src/sc_neurocore/accel/julia/neurons/ltc.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia for ltc

module LtcAccel

export step!, simulate, LiquidTimeConstantNeuronState

mutable struct LiquidTimeConstantNeuronState
    x::Float64
    tau_base::Float64
    w_tau::Float64
    w_x::Float64
    w_in::Float64
    bias::Float64
    v_threshold::Float64
    dt::Float64
end

function LiquidTimeConstantNeuronState()
    LiquidTimeConstantNeuronState(0.0, 10.0, -0.5, 0.8, 1.0, 0.0, 1.0, 1.0)
end

function step!(s::LiquidTimeConstantNeuronState, I_ext::Float64=0.0; dt::Float64=0.1)
    try
        tau = s.tau_base * (1.0 / (1.0 + exp(-(s.w_tau * I_ext + s.bias))))
        tau = max(tau, 0.1)
        f_target = tanh(s.w_x * s.x + s.w_in * I_ext)
        s.x += s.dt / tau * (-s.x + f_target)
        if s.x >= s.v_threshold
            s.x = 0.0
            return 1
        end
        return 0
    catch _e
        return 0
    end
end

function simulate(n_steps::Int=1000; I_ext::Float64=10.0, dt::Float64=0.1)
    s = LiquidTimeConstantNeuronState()
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, I_ext; dt=dt)
        trace[t] = s.x
        if result isa Number && result > 0
            spikes += 1
        end
    end
    return trace, spikes
end

end # module LtcAccel
